File: src/main/java/dev/slne/spawn/trader/command/subcommand/SpawnTraderSetCooldownCommand.java
package dev.slne.spawn.trader.command.subcommand;

import dev.jorel.commandapi.CommandAPI;
import dev.jorel.commandapi.CommandAPICommand;
import dev.jorel.commandapi.arguments.Argument;
import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.arguments.CustomArgument;
import dev.jorel.commandapi.arguments.LongArgument;
import dev.jorel.commandapi.arguments.PlayerArgument;
import dev.jorel.commandapi.arguments.StringArgument;
import dev.slne.spawn.trader.SpawnTrader;
import dev.slne.spawn.trader.manager.TradeManager;
import dev.slne.spawn.trader.manager.object.CooldownPair;
import dev.slne.spawn.trader.manager.object.Trade;
import dev.slne.spawn.trader.manager.object.impl.FrameTrade;
import dev.slne.spawn.trader.manager.object.impl.LightTrade;
import it.unimi.dsi.fastutil.objects.ObjectArrayList;
import it.unimi.dsi.fastutil.objects.ObjectList;
import java.util.UUID;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

/**
 * The type Spawn trader set cooldown command.
 */
public class SpawnTraderSetCooldownCommand extends CommandAPICommand {

  private final TradeManager tradeManager = SpawnTrader.instance().tradeManager();
  private final ObjectList<String> availableTrades = new ObjectArrayList<>();

  /**
   * Instantiates a new Spawn trader set cooldown command.
   *
   * @param name the name
   */
  public SpawnTraderSetCooldownCommand(String name) {
    super(name);

    withArguments(new PlayerArgument("target"));
    withArguments(tradeArgument().replaceSuggestions(
        ArgumentSuggestions.strings("light-block", "invisible-item-frame")));
    withArguments(new LongArgument("amount"));

    executesPlayer((player, args) -> {
      final Player target = Bukkit.getPlayer(
          args.getOrDefaultUnchecked("target", player.getName()));
      final long amount = args.getOrDefaultUnchecked("amount",
          SpawnTrader.instance().tradeCooldown());

      Trade trade = args.getUnchecked("trade");

      availableTrades.add(new LightTrade().name());
      availableTrades.add(new FrameTrade().name());

      if (!availableTrades.contains(trade.name())) {
        throw CommandAPI.failWithString("Der Trade wurde nicht gefunden!");
      }

      if (target == null) {
        player.sendMessage(SpawnTrader.prefix()
            .append(Component.text("Der Spieler wurde nicht gefunden.").color(NamedTextColor.RED)));
        return;
      }

      final UUID uuid = target.getUniqueId();
      CooldownPair cooldownPair = tradeManager.cooldownStorage()
          .getOrDefault(uuid, new CooldownPair(0L, 0L));

      final long currentTime = System.currentTimeMillis();
      final long cooldownEndTime = currentTime + amount;

      if (trade instanceof FrameTrade) {
        cooldownPair = new CooldownPair(cooldownEndTime, cooldownPair.getTrade1());
      } else if (trade instanceof LightTrade) {
        cooldownPair = new CooldownPair(cooldownPair.getTrade0(), cooldownEndTime);
      }

      tradeManager.cooldownStorage().put(uuid, cooldownPair);
      player.sendMessage(SpawnTrader.prefix().append(
          Component.text("Der Cooldown f\u00FCr den Trade wurde erfolgreich neu gesetzt.")));
    });
  }

  private Argument<Trade> tradeArgument() {
    return new CustomArgument<>(new StringArgument("trade"), info -> {
      Trade trade = Trade.getTrade(info.input());

      if (trade == null) {
        throw CustomArgument.CustomArgumentException.fromAdventureComponent(
            Component.text("Der Trade wurde nicht gefunden.").color(NamedTextColor.RED));
      } else {
        return trade;
      }
    });
  }
}

File: src/main/java/dev/slne/spawn/trader/command/subcommand/SpawnTraderSetCooldownCommand.kt
package dev.slne.spawn.trader.command.subcommand

import dev.jorel.commandapi.CommandAPICommand
import dev.jorel.commandapi.arguments.Argument
import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.CustomArgument
import dev.jorel.commandapi.arguments.StringArgument
import dev.jorel.commandapi.kotlindsl.longArgument
import dev.jorel.commandapi.kotlindsl.playerArgument
import dev.jorel.commandapi.kotlindsl.playerExecutor
import dev.slne.spawn.trader.manager.trade.Trade
import dev.slne.spawn.trader.manager.trade.impl.FrameTrade
import dev.slne.spawn.trader.manager.trade.impl.GlobeTrade
import dev.slne.spawn.trader.manager.trade.impl.LightTrade
import dev.slne.spawn.trader.plugin
import dev.slne.spawn.trader.util.TraderPermissionRegistry
import dev.slne.surf.surfapi.core.api.messages.adventure.buildText
import dev.slne.surf.surfapi.core.api.messages.adventure.sendText
import dev.slne.surf.surfapi.core.api.util.mutableObjectListOf
import org.bukkit.command.CommandSender

class SpawnTraderSetCooldownCommand(name: String) : CommandAPICommand(name) {
    private val availableTrades = mutableObjectListOf<String>()

    init {
        withPermission(TraderPermissionRegistry.COMMAND_SET_COOLDOWN)
        playerArgument("target")
        withArguments(
            tradeArgument().replaceSuggestions(
                ArgumentSuggestions.strings<CommandSender?>(
                    "light-block",
                    "invisible-item-frame",
                    "globe-banner-pattern"
                )
            )
        )
        longArgument("amount")

        availableTrades.add(LightTrade().name())
        availableTrades.add(FrameTrade().name())
        availableTrades.add(GlobeTrade().name())

        playerExecutor { player, args ->
            val target = args.getOrDefaultUnchecked("target", player)
            val amount = args.getUnchecked<Long>("amount") ?: return@playerExecutor
            val trade = args.getUnchecked<Trade>("trade") ?: return@playerExecutor

            val currentTime = System.currentTimeMillis()
            val cooldownEndTime = currentTime + amount

            plugin.saveCooldown(target, trade, cooldownEndTime)
            player.sendText {
                appendPrefix()
                success("Der Cooldown für den Trade ")
                variableValue(trade.name())
                success(" wurde für den Spieler ")
                variableValue(target.name)
                success(" auf ")
                variableValue(amount.toString())
                success(" Millisekunden gesetzt.")
            }
        }
    }

    private fun tradeArgument(): Argument<Trade> {
        return CustomArgument(
            StringArgument("trade")
        ) { info: CustomArgument.CustomArgumentInfo<String> ->
            val trade = Trade.getTrade(info.input())
                ?: throw CustomArgument.CustomArgumentException.fromAdventureComponent(
                    buildText {
                        appendPrefix()
                        error("Der Trade wurde nicht gefunden.")
                    }
                )

            trade
        }.replaceSuggestions(
            ArgumentSuggestions.strings {
                availableTrades.toTypedArray()
            }
        )
    }
}
